e2e/framework/case.go

```diff
@@ -0,0 +1,112 @@
+package framework
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestSuite defines a set of test cases and under what conditions to run them
+type TestSuite struct {
+	Component string // Name of the component/system/feature tested
+
+	CanRunLocal bool        // Flags if the cases are safe to run on a local nomad cluster
+	Cases       []TestCase  // Cases to run
+	Constraints Constraints // Environment constraints to follow
+	Parallel    bool        // If true, will run test cases in parallel
+	Slow        bool        // Slow test suites don't run by default
+
+	// API Clients
+	Consul bool
+	Vault  bool
+}
+
+// Constraints that must be satisfied for a TestSuite to run
+type Constraints struct {
+	Provider    string   // Cloud provider ex. 'aws', 'azure', 'gcp'
+	OS          string   // Operating system ex. 'windows', 'linux'
+	Arch        string   // CPU architecture ex. 'amd64', 'arm64'
+	Environment string   // Environment name ex. 'simple'
+	Tags        []string // Generic tags that must all exist in the environment
+}
+
+func (c Constraints) matches(env Environment) error {
+	if len(c.Provider) != 0 && c.Provider != env.Provider {
+		return fmt.Errorf("provider constraint does not match environment")
+	}
+
+	if len(c.OS) != 0 && c.OS != env.OS {
+		return fmt.Errorf("os constraint does not match environment")
+	}
+
+	if len(c.Arch) != 0 && c.Arch != env.Arch {
+		return fmt.Errorf("arch constraint does not match environment")
+	}
+
+	if len(c.Environment) != 0 && c.Environment != env.Name {
+		return fmt.Errorf("environment constraint does not match environment name")
+	}
+
+	for _, t := range c.Tags {
+		if _, ok := env.Tags[t]; !ok {
+			return fmt.Errorf("tags constraint failed, tag '%s' is not included in environment", t)
+		}
+	}
+	return nil
+}
+
+// TC is the base test case which should be embedded in TestCase implementations.
+// It also embeds testify assertions configured with the current *testing.T
+// context. For more information on assertions:
+// https://godoc.org/github.com/stretchr/testify/assert#Assertions
+type TC struct {
+	*assert.Assertions
```

Comment:
The reason for embedding assertions instead of require is so that any defined cleanup logic still runs in an `AfterEach` or `AfterAll` function.

Maybe we can check after we've run the test function if the testing.T context is marked as failed or not and do a hard fail after the Cleanup logic happens?

Reply:
> The reason for embedding assertions instead of require is so that any defined cleanup logic still runs in an AfterEach or AfterAll function.
Not sure I follow this argument. If the AfterEach/AfterAll run in defer blocks (as I see in the implementation), then why does it matter whether we embed assert or require? IIUC, defer blocks execute up the call chain even if there's a panic (the first non satisfied require line should cause a panic)

Reply:
Wrote a quick mock of verifying defer execution here https://play.golang.org/p/JYR6984Utkk

Reply:
So I actually agree with your previous suggesting of embedding `require.Assertions`. `t.FailNow` does an os.Exit which I thought didn't do deferred execution but from the [testing godoc](https://golang.org/pkg/testing/#T.FailNow) I read:
 > FailNow marks the function as having failed and stops its execution by calling runtime.Goexit (which then runs all deferred calls in the current goroutine). Execution will continue at the next test or benchmark.

Reply:
That's good to clarify. For context - I somehow landed on https://github.com/hashicorp/nomad/blob/c1fff02ec98b39ddcf6378a238748d4c5b3128eb/vendor/github.com/mitchellh/go-testing-interface/testing.go#L65 when clicking around the code, so I thought we were using this library everywhere and would see a panic. We do import that in a few places, but its not widespread.  t.FailNow does the right thing like you said so we should be fine.